docs: restructure working PR (#4866)
dfinity/motoko: doc/md/fundamentals/2-types/7-variants.md

---
sidebar_position: 7
---

# Variants

> Variants allow defining values that can take on multiple possible forms, each labeled with a distinct tag. Unlike [records](https://internetcomputer.org/docs/motoko/fundamentals/types/records), where all fields exist at once, a variant holds exactly one of its possible values at any given time. This makes variants useful for representing states, actions, or categorized data.

Reviewer: ```suggestion
Variants allow defining values that can take on multiple possible forms, each labeled with a distinct tag. Unlike [records](https://internetcomputer.org/docs/motoko/fundamentals/types/records), where all fields exist at once, a variant holds exactly one of its possible values at any given time. This makes variants useful for representing states, actions, or categorized data.
```
```suggestion
Variants allow defining values that take on one of several types, each labeled with a distinct  tag. Unlike [records](https://internetcomputer.org/docs/motoko/fundamentals/types/records), where all fields exist at once, a value of a variant type holds exactly one of the types' possible values. This makes variants useful for representing mutually exclusive alternatives such as states, enumerations, categories or taxonomies.
```

## Defining a variant

```motoko no-repl
type Status = {
    #Active;
    #Inactive;
    #Banned : Text;
};
```

> `#Active` and `#Inactive` are unit tags, meaning they do not store any extra data `()`. `#Banned` carries a [`Text`](https://internetcomputer.org/docs/motoko/base/Text) value, such as the reason for the ban.

Reviewer: ```suggestion
`#Active` and `#Inactive` are constant tags, with an implicit `()` argument, meaning they only store trivial data. `#Banned` carries a [`Text`](https://internetcomputer.org/docs/motoko/base/Text) value, such as the reason for the ban.
```

## Assigning variants

To assign a variant value, use one of the defined tags.

> ```motoko no-repl
> let activeUser : Status = #Active;
> let bannedUser : Status = #Banned("Violation of rules");
> ```

Reviewer: ````suggestion
The type annotations can also be omitted:

```motoko no-repl
let activeUser  = #Active;
let bannedUser = #Banned("Violation of rules");
```
````
